dmlc/xgboost
# [jvm-packages] [breaking] rework xgboost4j-spark and xgboost4j-spark-gpu

# About this PR

## Features and Fixes

- Introduce an abstract XGBoost Estimator:
  - Create an abstract base class that encapsulates common functionality for all XGBoost estimators.
  - This will promote code reuse and maintain consistency across different estimators.

- Update to the latest XGBoost parameters
  - Add all XGBoost parameters supported in XGBoost4j-spark base on [this](https://xgboost.readthedocs.io/en/stable/parameter.html)
  - Add setter and getter for these parameters.
  - Remove the deprecated parameters

- Address the missing value handling:
  - 0 is a meaningful value in machine learning case, the old version enforces 0 as the missing value which may result in XGBoost model inaccurate.

- Remove any ETL operations in XGBoost
  - As a thin wrapper of XGBoost4j, it shouldn't wrap much ETL in XGBoost4j itself to make code hard to maintain and read.

- Rework the GPU plugin:
  - Rework the GPU plugin to improve its stability and make it more readable and maintainable according to the new design pattern of xgboost4j-spark

- Expand sanity tests for CPU and GPU consistency:
  - Increase the coverage of sanity tests to verify that the training and transform results are consistent between CPU and GPU implementations.
  - This will help identify and prevent any discrepancies or bugs related to the GPU support.


## Breaking changes

- Remove some unused parameters like Rabit related parameters and train_test_ratio. etc.
- Separate XGBoost-spark parameter from the whole XGBoost parameters, and make the constructor of XGBoost estimator only accept the XGBoost parameters instead of parameters for xgboost4j-spark. Eg,
  ``` scala
  val xgbParams: Map[String, Any] = Map(
        "max_depth" -> 5,
        "eta" -> 0.2,
        "objective" -> "binary:logistic"
      )
  val est = new XGBoostClassifier(xgbParams)
        .setNumRound(10)
        .setDevice("cpu")
        .setNumWorkers(2)
  ``` 
   or

    ``` scala
    val est = new XGBoostClassifier(xgbParams)
          .setMaxDepth(5)
          .setEta(0.2)
          .setObjective("binary:logistic")
          .setNumRound(10)
          .setDevice("cpu")
          .setNumWorkers(2)
    ```

## Test this PR

- Unit tests For scala 2.12

``` bash
mvn clean package
mvn -P gpu clean package
```

- Unit tests For scala 2.13
``` bash
cd xgboost
python dev/change_scala_version.py --scala-version 2.13

cd jvm-packages
mvn clean package
mvn -P gpu clean package
```
